Hide PRB when all variations are out-of-stock

lovo-h opened the pull request:
## Changes proposed in this Pull Request:

Fixes #1580 

* Don't display the payment request button (PRB) when all variations of the product are out-of-stock.
* Don't display the PRB when an out-of-stock variation is selected.

This PR ensures that the PRB is not displayed when all of a product's variations are out of stock. Additionally, when a product has at least one variation that is not out-of-stock, this PR also ensures that the PRB is not displayed when the user selects a variation that is out-of-stock. 

This fixes an issue where users were seeing an "Out of stock" warning and the PRB, at the same time. Which may have caused some confusion. Now, only the warning is displayed. Also, this fix makes it so that the PRB's behavior in variable product pages matches that of the simple product pages.

At the moment, this entire part of the code is not being tested. Additionally, this change was a small enough not to warrant e2e tests.

<!--
Description: Write a summary of this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

**Don't display the PRB when all variations of a `variable product` are out-of-stock**
To test this new behavior, we need to create a variable product with at least two variants:

* Create a new product: `Products > Add New`
* Select "Variable product" as `Product data`
* Click `Attributes`, then click `Add` while "Custom product attribute" is selected
* Create "Size" attribute with "Small | Medium" as values
* Ensure `Used for variations` is checked
* Click `Variations` then while "Create variations from all attributes" is selected click `Go`
* Add a `Regular price ($)` for each variation
* Select "Out of stock" on the `Stock status` field, for each variation
* `Save changes` and go to the product's page

Note that the PRB is not visible. Also, note that switching between `Size` variations keeps the PRB hidden.

**Don't display the PRB when an out-of-stock variation is selected, for a `variable product`**
To test this new behavior, go back and edit the variable product created above and do the following:

* Click `Variations`
* Update the `Stock status` of the "small" variation to "In stock"
* `Save changes` and go to the product's page

Note that the PRB is visible, upon loading the product page. Also, note that selecting the "Small" `Size` variation makes the PRB show. But switching to the "Medium" variation makes the PRB hide. 

---

Note: for the following tests, the `WooCommerce Subscriptions` plugin will need to be active.

**Don't display the PRB when all variations of a `variable subscription` are out-of-stock**
To test this new behavior, we need to create a variable subscription with at least two variants:

* Create a new product: `Products > Add New`
* Select "Variable subscription" as `Product data`
* Click `Attributes`, then click `Add` while "Custom product attribute" is selected
* Create "Billing Period" attribute with "1 Month | 2 Months" as values
* Ensure `Used for variations` is checked
* Click `Variations` then while "Create variations from all attributes" is selected click `Go`
* Add a `Subscription price ($)` for each variation
* Select "Out of stock" on the `Stock status` field, for each variation
* `Save changes` and go to the product's page

Note that the PRB is not visible. Also, note that switching between "Billing Period" variations keeps the PRB hidden.

**Don't display the PRB when an out-of-stock variation is selected, for a `variable subscription`**
To test this new behavior, go back and edit the variable subscription created above and do the following:

* Click `Variations`
* Set the `Stock status` of the "1 Month" variation to "In stock"
* `Save changes` and go to the product's page

Note that the PRB is visible, upon loading the product page. Also, note that selecting the "1 Month" `Billing Period` variation makes the PRB show. But switching to the "2 Months" variation makes the PRB hide. 

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [x] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

lovo-h committed:
Hide "Buy Now" button when all variations are out of stock

Before, the "Buy Now" button would appear even if all variations of a
product were out-of-stock. This behavior did not match the one that
existed in the simple product pages.

Now, the "Buy Now" button only appears if there exists at least one
variation that is not out-of-stock. This fixes the mismatching behavior.

Fixes #1580

lovo-h committed:
Hide "Buy Now" button if a given variation is out-of-stock

Before, the "Buy Now" button would be displayed, even if the variation
was out-of-stock. This gave the false impression that the variation
could still be purchased.

Now, the "Buy Now" button is hidden when the user selects a variation
that is out-of-stock.

lovo-h commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -906,6 +906,32 @@ public function should_show_payment_request_button() {
 			return false;
 		}
 
+		if ( $this->is_product() && $this->get_product()->get_type() === 'variable' ) {
```
@asumaran, should we also apply this behavior to `variable-subscription`?

asumaran replied:
@lovo-h we should check if the fix also applies for variable subscriptions as well since we support variable subscriptions with PRBs https://github.com/woocommerce/woocommerce-gateway-stripe/blob/aa2de49012b8a434dd74e56742ef190a8fa4901c/includes/payment-methods/class-wc-stripe-payment-request.php#L560-L562

lovo-h replied:
I tested for this and it looks like this fix should also be applied to `variable-subscription`. 

I'll go ahead and update this piece of code and update the PR's notes so that it includes a test for `variable-subscription`.

lovo-h committed:
Add changelog entry: Hide PRB when variations are out-of-stock

asumaran commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -906,6 +906,32 @@ public function should_show_payment_request_button() {
+			];
+
+			// Retrieve variable product's variations that are not 'outofstock' (e.g. 'instock', 'onbackorder').
+			$variations = wc_get_products( $product_args );
```
If I read this correctly. Here we are trying to hide the button if **all** variations are out of stock right? We could leverage `$this->get_product()->get_available_variations()` like this:

```php
if ( $this->is_product() && $this->get_product()->get_type() === 'variable' ) {
  $stock_availability = array_column($this->get_product()->get_available_variations(), 'is_in_stock');
  return array_unique($stock_availability) !== [false];
}
```

Which will hide the PRB if all variations are out of stock.

lovo-h replied:
Pretty neat. Didn't know we already had `get_available_variations()`. Good call 👍

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -649,7 +653,11 @@ jQuery( function( $ ) {
+						if (response.is_in_stock) {
+							$( document.body ).trigger( 'wc_stripe_unblock_payment_request_button' );
+						} else {
+							$( document.body ).trigger( 'wc_stripe_hide_payment_request_button' );
```
I think it's better if we use `wc_stripe_payment_request.unblockPaymentRequestButton()` directly here instead of registering and firing a custom jQuery event. In the case if `wc_stripe_unblock_payment_request_button` we created an event with that name because we wanted 3rd party extensions to be able to trigger the event.

lovo-h replied:
Ah, I see. Good call.

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -675,7 +683,11 @@ jQuery( function( $ ) {
+							if (response.is_in_stock) {
+								$( document.body ).trigger( 'wc_stripe_unblock_payment_request_button' );
+							} else {
+								$( document.body ).trigger( 'wc_stripe_hide_payment_request_button' );
```
same as previous comment.

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -675,7 +683,11 @@ jQuery( function( $ ) {
 							} )
 						).then( function () {
-							$( document.body ).trigger( 'wc_stripe_unblock_payment_request_button' );
+							if (response.is_in_stock) {
```
Not sure waiting for the backend response to show/hide the PRB is best here. I mean, it'd be the best option if we also updated the whole product data in the frontend (which we currently do not). I think we should get the `is_in_stock` for each variant from the form (`$('.variations_form').data('product_variations')[0].is_in_stock;`) and show/hide the button accordingly.

lovo-h replied:
That's true: waiting until the backend responds to update the PRB does seem a bit odd. Will update 👍

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -718,6 +730,20 @@ jQuery( function( $ ) {
+		},
+
+		unhidePaymentRequestButton: function () {
+			const stripe_wrapper_ele = $( '#wc-stripe-payment-request-wrapper' );
+			const stripe_separator_ele = $( '#wc-stripe-payment-request-button-separator' );
```
no need to add `_ele` suffix

asumaran reviewed (changes requested):
PR is looking good! I'm glad that you were able to work on this issue on your own. I left a few comments for you @lovo-h. Let me know if you have any questions. Happy to discuss any of the comments I'm leaving.

lovo-h committed:
Hide "variable-subscription" if all variations out-of-stock

Applying the same rule to hide "variable-subscription", like the one
that is applied to "variable" products.

Before, the payment request button (PRB) would be hidden if all
variations of a "variable" product were out of stock but not if all
variations of a "variable-subscription" product. This was inconsistent
behavior.

lovo-h committed:
Update logic that determines if all variations are out-of-stock

Leveraging an existing method (e.g. 'get_available_variations()') to
retrieve all of the variations and parse out whether there's at least
one "in-stock".

The tradeoff here is that 'get_available_variations()' makes multiple
calls to the DB (through 'wc_get_product($variation_id)') in order to
retrieve the data of each variation. But at the same time, we obtain
more readable, hence, more maintainable code.

lovo-h committed:
Immediately hide/display PRB, based on whether it's in-stock

Before, the payment request button (PRB) would wait for a backend
request to respond before hiding/displaying the PRB. This created a bit
of an odd behavior.

Now, the PRB is displayed/hidden based on the product's cached-data.

lovo-h committed:
Rename variables to remove unnecessary suffix (e.g. "_ele")

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -453,6 +453,53 @@ jQuery( function( $ ) {
 			} );
 		},
 
+		getSelectedVariationCachedData: function () {
```
I think we should minimize introducing code outside the scope of Payment Request buttons in this file and in the extension in general.

Note that when changing the dropdowns the product changes according to the selected attributes. So there must be code that can get the right variation.

<img src="https://user-images.githubusercontent.com/1025173/141597485-98355633-6789-4845-a072-1bbedc90d403.gif" width="320" />

And that code is here: 
 https://github.com/woocommerce/woocommerce/blob/fd3dfdbf3e05ce87138831d2f8eff3734eeb4572/plugins/woocommerce/assets/js/frontend/add-to-cart-variation.js#L239

Which is called here if the variation is found https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/assets/js/frontend/add-to-cart-variation.js#L210 

I think we should leverage the existing events attached to the form and use it to get the selected variation and with that we could hide/unhide the PRB accordingly.

The code can be simplified to just (within `attachProductPageEventListeners()` method):


```js
if ( $('.variations_form').length ) {
  $('.variations_form').on('found_variation.wc-variation-form', function (evt, variation) {
    if (variation.is_in_stock) {
      wc_stripe_payment_request.unhidePaymentRequestButton();
    } else {
      wc_stripe_payment_request.hidePaymentRequestButton();
    }
  });
}
```

lovo-h replied:
I didn't notice that there was existing code that already retrieved the right variation 🤦  

Good call. That definitely looks cleaner 👍

asumaran reviewed (changes requested):
@lovo-h Unfortunately, I think we need to do some changes on how do we get the selected variation.

lovo-h committed:
Leverage existing event for retrieving variation, on change

Before, the code duplicated existing functionality that retrieved some
selected variation.

Now, the code leverages the existing functionality by subscribing to the
event that gets triggered when some valid variation is selected.

asumaran reviewed (approved):
LGTM! :shipit: 

#### Variable Products

- [x] 2 variation (color:red|green) in stock,
  - [x] no default - PRB is displayed but can't open payment sheet
  - [x] red is selected - PRB is displayed, payment sheet opens
  - [x] green is selected - PRB is displayed, payment sheet opens, price is updated
  - [x] red and green default - PRB is displayed, payment sheet opens, price is updated when changing variations
- [x] 2 variation (color:red|green) green is out of stock,
  - [x] no default - PRB is displayed but can't open payment sheet
  - [x] red is selected - PRB is displayed
  - [x] green is selected - PRB hides
  - [x] red as default - PRB is displayed, payment sheet opens - switching to green PRB hides
  - [x] green as default - PRB is displayed but can't open payment sheet
- [x] 4 variation (color: red|green, size: small|medium) - red-medium and green-medium are out of stock
  - [x] no default - PRB is displayed but can't open payment sheet
  - [x] red-small is selected - PRB is displayed, payment sheet opens
  - [x] red-medium is selected - PRB hides
  - [x] red-small/green-small as default - PRB is displayed, payment sheet opens - switching to red-medium PRB hides
  - [x] red-medium/green-medium as default - PRB is displayed but can't open payment sheet

#### Variable Subscriptions

- [x] 4 variation (color: red|green, size: small|medium) - red-medium and green-medium are out of stock - logged out user
  - [x] no defaults - works as expected with all cases - displays alert user must login.
  - [x] with defaults - works as expected
- [x] 4 variation (color: red|green, size: small|medium) - red-medium and green-medium are out of stock - logged in user
  - [x] no defaults - works as expected
  - [x] with defaults - works as expected
